File: Leetcode/Tree/GFG Root to Leaf Paths.cpp
#include <iostream>
#include <bits/stdc++.h>

using namespace std;

struct Node
{
    int data;
    struct Node* left;
    struct Node* right;

    Node(int x)
    {
        data = x;
        left = right = NULL;
    }
};


class Solution 
{
  public:
    
    void dfs(Node* node, vector<vector<int>>& ans, vector<int>& path)
    {
        if(node==NULL)
            return;
        
        path.push_back(node->data);
        
        if(node->left == NULL && node->right == NULL)
        {
            ans.push_back(path);
        }
        
        dfs(node->left, ans, path);
        
        dfs(node->right, ans, path);
        
        //IMP - Backtracking step
        path.pop_back();
    }
    
    vector<vector<int>> Paths(Node* root) 
    {
        
        vector<vector<int>> ans;
        vector<int> path;
        
        dfs(root, ans, path);
        
        return ans;
        
    }
};

File: Leetcode/Tree/GFG Root to Leaf Paths.cs
using System;
using System.Collections.Generic;

public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;
    
    public TreeNode(int x) { val = x; }
}

public class Solution
{
    private void dfs(TreeNode node, List<List<int>> ans, List<int> path)
    {
        if (node == null)
            return;

        path.Add(node.val);

        if (node.left == null && node.right == null)
        {
            ans.Add(new List<int>(path)); // Make a DEEP Copy of the path list
        }
        
        dfs(node.left, ans, path);
        dfs(node.right, ans, path);

        // IMP Backtrack: remove the current node from path (last node in path)
        path.RemoveAt(path.Count - 1);
    }


    public List<List<int>> Paths(TreeNode root)
    {
        List<List<int>> ans = new List<List<int>>();
        List<int> path = new List<int>();
        
        // Start DFS from the root node
        dfs(root, ans, path);
        
        return ans;
    }
}

class Program
{
    static void Main()
    {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        Solution solution = new Solution();
        List<List<int>> result = solution.Paths(root);

        foreach (var path in result)
        {
            Console.WriteLine(string.Join(" ", path));
        }
    }
}
